flang/lib/Lower/FIRBuilder.cpp

```diff
@@ -166,21 +166,40 @@ struct CharacterOpsBuilderImpl {
       return fir::ReferenceType::get(getCharacterType());
     }
 
-    bool isConstant() const { return data.getType().isa<fir::SequenceType>(); }
+    bool isNotInMemory() const {
+      return data.getType().isa<fir::SequenceType>() ||
+             data.getType().isa<fir::CharacterType>();
+    }
+
+    std::optional<fir::SequenceType::Extent> getCompileTimeLenght() const {
+      auto type = data.getType();
+      if (type.isa<fir::CharacterType>())
+        return 1;
+      if (auto refType = type.dyn_cast<fir::ReferenceType>())
+        type = refType.getEleTy();
+      if (auto seqType = type.dyn_cast<fir::SequenceType>()) {
+        auto shape = seqType.getShape();
+        assert(shape.size() == 1 && "only scalar character supported");
+        return shape[0];
+      }
+      return {};
+    }
 
     /// Data must be of type:
     /// - fir.ref<fir.char<kind>> (dynamic length)
     /// - fir.ref<fir.array<len x fir.char<kind>>> (len compile time constant).
     /// - fir.array<len x fir.char<kind>> (character constant)
+    /// - fir.char<kind> (length one character)
     mlir::Value data;
     mlir::Value len;
   };
 
-  Char materializeConstant(Char cst) {
-    assert(cst.isConstant() && "expected constant");
-    auto variable = builder.createHere<fir::AllocaOp>(cst.data.getType());
-    builder.createHere<fir::StoreOp>(cst.data, variable);
-    return {variable, cst.len};
+  Char materializeValue(Char str) {
+    if (!str.isNotInMemory())
```

Comment:
Right. Consider maybe renaming the test to something like `needToMaterialize` perhaps?